### DISCUSSION TITLE
Smooth scrolling

### DISCUSSION THREAD

**Discussion by RGBCube** [ghostty-org/ghostty#2355] [Feature Requests, Ideas]
As the title says, smooth scrolling (bonus points if the beizer curve is configurable) would be nice

**Comment by mitchellh** [comment]
Thinking about it.

**Comment by RGBCube** [comment]
Idea: Only scroll the lines that change, so the bottom bar in Vim or any other editor doesn't flicker. Would make a much smoother experience

**Comment by mitchellh** [comment]
Closed #1934 for this, really "merged" then. As we consider smooth scrolling we can address when the cursor is part of the scroll and how we can perhaps bring generalized animation into our renderers.

**Comment by mitchellh** [comment]
Converted this to a discussion because it isn't directly actionable and a significant amount of design work needs to be done. We can tear off actionable issues as we get there.

**Comment by sid314** [comment]
Getting smooth scroll working in programs like Neovim will be more difficult. For example take Neovide. In Neovide if you open a terminal inside Neovim, the smooth scroll dosent work inside that. That is bexause Neovim does its own thing inside that terminal and does not know that it is running inside Neovide.

**Reply by theherk** [reply]
My terminal in Neovide is smooth scrolling; both scrolling and long-running output. However, neovim inside a terminal running in neovide is again not smooth scrolling for the reason you point out. There needs to be knowledge about scrollable regions. But the terminal in Neovide is buttery smooth.

Even smoother than this video shows.
https://github.com/user-attachments/assets/f19c10b7-1deb-4795-953d-735b3d728864

**Reply by sid314** [reply]
Yes that is what i meant.

**Comment by sweiglbosker** [comment]
Ghostty is a terminal emulator, not a neovim client. This would be impossible to implement because there is no notion of partial scrolling in the terminal, just the redrawing of lines, cursor movement, and whatever else is described by ansi escape codes these days. Any attempt at adding smooth scrolling would only be as capable as something like tmux copy-mode.

**Reply by 00-kat** [reply]
Smooth scrolling of the scrollback in the primary screen is definitely possible.

**Reply by sweiglbosker** [reply]
Yes, that's what I was trying to point out with the tmux copy mode analogy. The impossible part I mention (at least without writing a vim client) is the feature brought up here https://github.com/ghostty-org/ghostty/discussions/2355#discussioncomment-10823417

**Reply by theherk** [reply]
I think you're right... for now. But until terminals like Ghostty, Contour, or Wezterm start implementing a partial scroll standard there will be no efforts made in TUI's or TUI libraries to support it. But I imagine, it will become a widely used feature someday (perhaps the distant future).

**Reply by 00-kat** [reply]
I'm not too sure how it works since I've never used it before, but “[Scroll Margins]” looks like an existing scroll standard… (Not sure if it works in alternate screen though.)

There's also an escape sequence to pan the primary screen, though that of course doesn't apply to the alternate screen.

[Scroll Margins]: https://vt100.net/docs/vt220-rm/chapter4.html#S4.13

**Reply by sweiglbosker** [reply]
Not sure if the dislikes are due to not liking that its impossible (same bro) or disagreement, but for those curious about what a protocol that allows this would like like you can read about how neovim’s rpc protocol handles scrolling [here](https://neovim.io/doc/user/ui.html#ui-event-highlight_set). Its also worth reading about [linegrid](https://neovim.io/doc/user/ui.html#ui-linegrid) and [multigrid](https://neovim.io/doc/user/ui.html#ui-multigrid).

**Reply by 00-kat** [reply]
I'm honestly not sure if it's *definitely* impossible to have heuristics, something like “when synchronised rendering is enabled, if a line[^1] is identical[^2] to another but is in another position, assume it was scrolled” would work I think.  I say synchronised rendering as a pre-requisite because it allows the terminal emulator to know when a frame has been fully rendered, so that incomplete renderings aren't smoothly scrolled.

Of course, heuristics does sound like a pain to implement, and probably would have *many* edge cases.

[^1]: Not sure how easy it would be to handle multiple buffers in the same window.  If it's not restricted to a line then it would be much more complicated.
[^2]: And a bit more complexity around when there's multiple identical lines.

**Reply by theherk** [reply]
I suspect the dislikes are due to you stating confidently that this is impossible when it isn't. That's why people like @mitchellh are talking about it at all, because while difficult, it isn't impossible. 

Yes the terminal doesn't have a pixel-wise concept. But like you said, this is a terminal emulator, and the thing wrapping the terminal can implement pixel-wise scrolling for a set of lines greater than those in the viewport. 

You can see this already in neovide or even an old vt100. The trouble is how to handle scrollable regions independently and that would probably be done with some control code for the program to instruct the emulator what scrolling candidates needed redraw.

Not only is it possible, I suspect one day it will be ubiquitous.

**Reply by 00-kat** [reply]
> You can see this already in neovide or even an old vt100.

Knowing absolutely nothing about them: VT100s didn't have alternate screens though, so did they handle smooth scrolling of editors like Vi fine?

**Reply by theherk** [reply]
Definitely not. To date there is no standard for defining scrollable regions. What it did have was pixel-wise scrolling of the viewport. Combining that with a standard for communicating scrollable regions is what will make that doable, I think.

**Reply by tonku321** [reply]
had to drop using ghostty because its not supported yet

**Reply by theherk** [reply]
> had to drop using ghostty because its not supported yet

That makes no sense as no terminals support it yet. If you know of one, please do share.

**Reply by sweiglbosker** [reply]
> I suspect the dislikes are due to you stating confidently that this is impossible when it isn't. That's why people like @mitchellh are talking about it at all, because while difficult, it isn't impossible.
> 
> Yes the terminal doesn't have a pixel-wise concept. But like you said, this is a terminal emulator, and the thing wrapping the terminal can implement pixel-wise scrolling for a set of lines greater than those in the viewport.
> 
> You can see this already in neovide or even an old vt100. The trouble is how to handle scrollable regions independently and that would probably be done with some control code for the program to instruct the emulator what scrolling candidates needed redraw.
> 
> Not only is it possible, I suspect one day it will be ubiquitous.

Sorry i realize my statements were a bit confusing. Im not saying pixel wise scrolling is impossible. In fact, literally every gtk vte terminal emulator supports it. Im saying that pixel wise scrolling in vim splits or tui apps is impossible for a terminal emulator to do in a straightforward and portable way.

**Reply by theherk** [reply]
Yeah I think I get your point, @sweiglbosker; I'm just disagreeing with it. As affably as possible, of course. I'm saying I see a future where Ghostty, Contour, Wezterm, and that ilk adopt, after its creation, a standard control code mechanism for scrolling regions in the viewport. From there, TUI's, will be able to send these instructions.

**Reply by sweiglbosker** [reply]
Yeah, the whole point of my argument was to say that we need a new protocol

**Reply by alicealysia** [reply]
except it literally has been implemented by a community patch already:
https://github.com/ghostty-org/ghostty/discussions/3206

**Reply by sweiglbosker** [reply]
Okay i think im done restating my point after this lol, but if you actually read my message, i am talking about content aware partial scrolling (in lieu of https://github.com/ghostty-org/ghostty/discussions/2355#discussioncomment-10823417) not pixelwise scrolling which is trivial to implement and support

**Reply by j4james** [reply]
@sweiglbosker Just FYI, the protocol for this does actually exist already. If an app wants to smoothly scroll an area of the screen, it can set the boundaries with [`DECSTBM`](https://vt100.net/docs/vt510-rm/DECSTBM.html) and [`DECSLRM`](https://vt100.net/docs/vt510-rm/DECSLRM.html), enable the smooth scrolling mode with [`DECSCLM`](https://vt100.net/docs/vt510-rm/DECSCLM.html), and then output a linefeed or [`IND`](https://vt100.net/docs/vt510-rm/IND.html) at the bottom of the margin area to trigger the scroll (or [`RI`](https://vt100.net/docs/vt510-rm/chapter4.html#T4-2) at the top of the margin area to scroll in the other direction).

When scrolling a large amount, and you don't want the delay of smooth scrolling, you just reset `DECSCLM`. You can even control the speed of the scrolling to some extent with the [`DECSSCLS`](https://vt100.net/docs/vt510-rm/DECSSCLS.html) setting. And on terminals that don't support this mode (or if it has been disabled), it'll just fallback to regular scrolling.

**Reply by alicealysia** [reply]
> @sweiglbosker Just FYI, the protocol for this does actually exist already. If an app wants to smoothly scroll an area of the screen, it can set the boundaries with [`DECSTBM`](https://vt100.net/docs/vt510-rm/DECSTBM.html) and [`DECSLRM`](https://vt100.net/docs/vt510-rm/DECSLRM.html), enable the smooth scrolling mode with [`DECSCLM`](https://vt100.net/docs/vt510-rm/DECSCLM.html), and then output a linefeed or [`IND`](https://vt100.net/docs/vt510-rm/IND.html) at the bottom of the margin area to trigger the scroll (or [`RI`](https://vt100.net/docs/vt510-rm/chapter4.html#T4-2) at the top of the margin area to scroll in the other direction).
> 
> When scrolling a large amount, and you don't want the delay of smooth scrolling, you just reset `DECSCLM`. You can even control the speed of the scrolling to some extent with the [`DECSSCLS`](https://vt100.net/docs/vt510-rm/DECSSCLS.html) setting. And on terminals that don't support this mode (or if it has been disabled), it'll just fallback to regular scrolling.

Is there any terminal emulators with all these controls present?

**Comment by pluiedev** [comment]
Let's just focus on scrolling on the primary screen for now - alternate screen smooth scrolling would likely require a separate protocol for defining partial scroll regions within TUIs, which is a much more complex task. 

On the GTK side I think this will come naturally with #111. Instead of handling mouse events ourselves we should rely on GTK's scrollbar system, and map scrollbar positions into lines. Then, all we need to do to implement smooth scrolling is to support the idea of "fractional lines" in the renderer. Everything after that should come smoothly (pun intended).

**Comment by tonku321** [comment]
I think it would be good if something like this could be implemented through some terminal API that could be called from, for example, the nvim extension. Then there is no need to store code to support different programs in the ghostty repo, and different programs can implement their own behavior for pixel based scrolling.

for example this function will work well for nvim only if ui elements (statusline, ...) are not scrolled

**Reply by pluiedev** [reply]
As I said, let's focus on primary screen (aka non-TUI) scrolling for now. Alternate scrolling is a much more convoluted topic that I think warrants its own discussion, especially since it requires collaboration with TUI authors.

**Reply by theherk** [reply]
I don't disagree that smooth scrolling the viewport should be worked on first, but I do disagree that region scrolling discussion should be tabled. It is important to have this discussion early because the implementations will no doubt be closely related, and measuring twice before making a cut seems prudent.

Also, it doesn't require close collaboration with TUI's to implement. They will simply be able to make use of the API once implemented.

**Reply by pluiedev** [reply]
I think it wouldn't be as similar as you would imagine - firstly, having separate regions would likely mean we need to somehow partition the screen and store individual scroll metadata for each region, which requires a lot more extra data keeping in Ghostty core compared to storing a global fractional scroll offset. On the renderer side we either require issuing separate draw calls for each region, or put scroll data as a per-instance attribute instead of a shader uniform, and both approaches are very costly and far more complicated than what we can get by with in a full scroll on the primary screen, which is again literally just two variables.

We do need to collaborate with TUIs to gather what the *use case* is and how the ergonomics would be on *their* side when using the API. You can't make an informed design on an API if you don't even know what it would be used for and *how* it would be used, and by the time you need to adjust the design to cater to downstream needs, you don't have much room to make backwards-compatible changes. A bottom-up (from concrete cases to abstractions) approach here IMO would be necessary.

**Reply by dwt** [reply]
Regarding smooth scrolling of regions, it seems that dec vt320 terminals could do this in hardware. I have found some videos of this
- [Dec vt320 smooth scrolling](https://www.youtube.com/watch?v=tSJfzrSA0ec)
- [Playing with a DEC VT320](https://www.youtube.com/watch?v=v_ol5ngrwgY)

I do not have much information about that, but from what little I found, the terminal actually detected if a scroll happened in hardware. The first video clearly shows it scrolling only a part of the screen, so this was not limited to full screen scrolling.

I have also [found this guy, who claims to have implemented this with custom controls sequences](https://flak.tedunangst.com/post/terminal-smooth-scrolling), but I'm not sure which terminal he is talking about.

**Reply by j4james** [reply]
@dwt The control sequence they're using is just the DEC scrolling mode ([`DECSCLM`](https://vt100.net/docs/vt510-rm/DECSCLM.html)), although it's incorrectly represented in that article. It's a level 1 conformance requirement, so it should be supported by all DEC terminals from the VT100 onwards. Ghostty actually claims to support it, but last time I checked it didn't seem to have any effect.

**Reply by alicealysia** [reply]
> Regarding smooth scrolling of regions, it seems that dec vt320 terminals could do this in hardware. I have found some videos of this
> 
>     * [Dec vt320 smooth scrolling](https://www.youtube.com/watch?v=tSJfzrSA0ec)
> 
>     * [Playing with a DEC VT320](https://www.youtube.com/watch?v=v_ol5ngrwgY)
> 
> 
> I do not have much information about that, but from what little I found, the terminal actually detected if a scroll happened in hardware. The first video clearly shows it scrolling only a part of the screen, so this was not limited to full screen scrolling.
> 
> I have also [found this guy, who claims to have implemented this with custom controls sequences](https://flak.tedunangst.com/post/terminal-smooth-scrolling), but I'm not sure which terminal he is talking about.

He did it in his own terminal emulator called vertigo. It prides itself on being the "worst terminal emulator"

**Comment by dwt** [comment]
As it doesn't seem to have been said in above discussion: Smooth scrolling for the main terminal window is really helpful, as it allows you to scroll fast and slow through huge volumes of text without loosing your reading position as easily.

This helps greatly with glancing through programs that generate a lot of output. For me, one prime example are docker builds, or ansible deployments. Especially the latter create huge output volumes in check mode that I need to quickly look through to see if there are unexpected changes.

And while of course that is possible with line scrolling (like it is implemented now) I like smooth scrolling much better and think it helps me with retaining context while scrolling as lines do not jump.

**Reply by sweiglbosker** [reply]
Agreed; for me things like animated cursor or smooth scrolling is not eye candy, but a QOL feature.

**Reply by leoauri** [reply]
Edit: apparently "smooth scrolling" here does not refer to pixel-by-pixel (as opposed to line-by-line) scrolling. The discussion around pixel-level scrolling is here: https://github.com/ghostty-org/ghostty/discussions/3206

**Comment by taysta** [comment]
The UX benefit of smooth scrolling is that it is easier to see how far you're scrolling, without a loss of context.

I'd like to add that VT320 terminals had it, and it looked amazing back then too: https://www.youtube.com/watch?v=tSJfzrSA0ec

I understand why this is probably very low priority, but myself and many other would love to see this feature!
